lib/auth/join_oracle.go

```diff
@@ -18,20 +18,224 @@ package auth
 
 import (
 	"context"
+	"encoding/base64"
+	"net/http"
+	"slices"
+	"strings"
+	"time"
 
 	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
 
 	"github.com/gravitational/teleport/api/client"
 	"github.com/gravitational/teleport/api/client/proto"
 	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/auth/join/oracle"
 )
 
+func generateOracleChallenge() (string, error) {
+	challenge, err := generateChallenge(base64.StdEncoding, 32)
+	return challenge, trace.Wrap(err)
+}
+
+func checkHeaders(headers http.Header, challenge string, clock clockwork.Clock) error {
+	// Check that required headers are signed.
+	authHeader := oracle.GetAuthorizationHeaderValues(headers)
+	rawSignedHeaders, ok := authHeader["headers"]
+	if !ok {
+		return trace.BadParameter("missing list of signed headers")
+	}
+	signedHeaders := strings.Split(rawSignedHeaders, " ")
+	if !slices.Contains(signedHeaders, oracle.DateHeader) {
+		return trace.BadParameter("header %s is not signed", oracle.DateHeader)
+	}
+	if !slices.Contains(signedHeaders, oracle.ChallengeHeader) {
+		return trace.BadParameter("header %s is not signed", oracle.ChallengeHeader)
+	}
+
+	// Check X-Date.
+	now := clock.Now().UTC()
+	rawHeaderDate := headers.Get(oracle.DateHeader)
+	if rawHeaderDate == "" {
+		return trace.BadParameter("missing header %v", oracle.DateHeader)
+	}
+	headerDate, err := time.Parse(http.TimeFormat, rawHeaderDate)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	var timeDelta time.Duration
+	if headerDate.After(now) {
+		timeDelta = headerDate.Sub(now)
+	} else {
+		timeDelta = now.Sub(headerDate)
+	}
+	if timeDelta > 5*time.Minute {
+		return trace.BadParameter("header time is more than 5 minutes from now")
+	}
```

Comment:
Is this supposed to prevent replay attacks, or to make sure we don't have clock drift? Because if it's about replay attacks, then perhaps we should probably just perform the check in one direction, and if it's about clock drift, then perhaps 5 minutes of tolerance is too much?